File: sugars/c++/eval_order.cpp
#include <iostream>
#include <type_traits>

using namespace std;


// Notice that these errors happen in order of their declaration:

void first(int arg) {
  static_assert(sizeof(int) == 7, "first");
}

void second() {
  static_assert(sizeof(int) == 7, "second");
}

// But these happen out of order of their declaration,
// because the static_asserts happen at expansion time:

template<typename B>
int outer(int arg) {
  static_assert(sizeof(B*) == 7, "Begin outer call");
  int answer = 0;
  static_assert(sizeof(B*) == 7, "End outer call");
  return answer;
}

template<typename A>
int inner() {
  static_assert(sizeof(A*) == 7, "Begin inner call");
  return 0;
  static_assert(sizeof(A*) == 7, "End inner call");
}

// And the order of the errors suggests that
// 'inner' is expanded before 'outer':

int main() {
  return outer<int>(inner<int>());
}

File: sugars/julia/eval_order.jl
macro outer(x)
    print("Begin outer call\n")
    answer = :($x * 2)
    print("End outer call\n")
    return answer
end

macro inner(x)
    print("Begin inner call\n")
    answer = :($x + 1)
    print("End inner call\n")
    return answer
end

print("(10 + 1) * 2 = ", @outer(@inner(10)), "\n")

#=
Output:
  Begin outer call
  End outer call
  Begin inner call
  End inner call
  (10 + 1) * 2 = 22
Evaluation order: OI*
Syntax safe: yes
=#